### ISSUE TITLE
Cannot extract dns_name from VPC Endpoint when using count syntax

### ISSUE THREAD

**Issue by jmoney8080** [hashicorp/terraform#20123]
### Terraform Version
```
Terraform v0.11.8
```

### Terraform Configuration Files
```hcl
data "aws_security_group" "internal" {
  id = "${var.security_group_id}"
}

data "aws_subnet_ids" "vpc_subnets" {
  vpc_id = "${data.aws_security_group.internal.vpc_id}"

  tags = {
      Purpose = "internal"
  }
}

resource "aws_vpc_endpoint" "vpce" {
    count               = "${length(data.aws_subnet_ids.vpc_subnets.ids)}"
    vpc_id              = "${data.aws_security_group.internal.vpc_id}"
    service_name        = "${var.vpce_service_name}"
    vpc_endpoint_type   = "Interface"
    security_group_ids  = ["${data.aws_security_group.internal.id}"]
    subnet_ids          = ["${element(data.aws_subnet_ids.vpc_subnets.ids, count.index)}"]
    private_dns_enabled = false
}


resource "aws_route53_record" "vpce-dns-record" {
    count   = "${length(data.aws_subnet_ids.vpc_subnets.ids)}"
    zone_id = "${var.r53_zone}"
    name    = "${var.application_service_name}.${var.aws_region}"
    type    = "CNAME"
    ttl     = "60"

    records = ["${lookup(element(element(aws_vpc_endpoint.vpce.*.dns_entry, count.index), 0), "dns_name")}"]

    weighted_routing_policy {
        weight = 10
    }

    set_identifier = "vpce-endpoint-${count.index}"
}
```

### Expected Behavior
Extracted the dns_name out of the output of the vpc endpoint list to apply to a DNS record.

### Actual Behavior
```
${lookup(element(element(aws_vpc_endpoint.vpce.*.dns_entry, count.index), 0), "dns_name")}
* module.attribute-content-service.aws_route53_record.vpce-dns-record[0]: At column 10, line 1: element: argument 1 should be type list, got type string in:
```

### Steps to Reproduce
1. `terraform init`
2. `terraform plan -out the_plan.tfplan`

### Additional Context
I've also tried `records = ["${lookup(aws_vpc_endpoint.vpce.*.dns_entry[count.index][0], "dns_name")}"]` hoping that would work from #11036 (comment) but a syntax error was thrown instead.

The above seems to fail as this example is `List<List<Map<String, String>>>`.  The inner `Map<String, String>` is the `dns_entry` output from the VPC Endpoint resource.  The wrapping `List` is that the VPC Endpoint has a list of `dns_entry`.  The final output `List` is that the VPC Endpoint resource is a `List` as the `count` attribute was used.

Alternatively, the aws-provider could export the dns entries as individual items instead of in the 1dns_entry1 map but that I believe would be a breaking change.

The above is my real world issue I'm trying to solve. I believe I've re-created this in a simpler [setup](https://github.com/jmoney8080/terraform-index-test).

### References
* https://github.com/jmoney8080/terraform-index-test

**Comment by jmoney8080** [comment]
I've also tried `records = ["${lookup(aws_vpc_endpoint.vpce.*.dns_entry[count.index][0], "dns_name")}"]` hoping that would work from https://github.com/hashicorp/terraform/issues/11036#issuecomment-357334325 but a syntax error was thrown instead.

**Comment by jmoney8080** [comment]
Alternatively, the aws-provider could export the dns entries as individual items instead of in the `dns_entry` map but that I believe would be a breaking change.

**Comment by jmoney8080** [comment]
The above is my real world issue I'm trying to solve.   I believe I've re-created this in a simpler setup.

https://github.com/jmoney8080/terraform-index-test

**Comment by mildwonkey** [comment]
Hi @jmoney8080,

I'm sorry you've run across this behavior, and thank you for the excellent reproduction setup. 
Much like the linked issue, this is a known limitation of the terraform configuration language and I can confirm that the same output will work as expected in terraform 0.12 (using your example setup):

```bash
$ terraform output
list_list_map = [
  [
    {
      "bar" = "baz"
      "foo" = "bar"
    },
  ],
]
value_from_inner_map_element = bar
```


For my own edification, I played around with `locals` until I could coax `baz` out of `list_list_map`. This might give you something to play with as a workaround, but I'm not sure if it's actually any use in the real world:

```hcl
locals {
  inner_list = "${var.list_list_map[0]}"
  inner_map = "${local.inner_list[0]}"
  lookup = "${lookup(local.inner_map, "bar")}"
}


output "value_from_inner_map_element" {
  value = "${local.lookup}"
}
```

```bash
$ terraform output
list_list_map = [
    [
        map[bar:baz foo:bar]
    ]
]
value_from_inner_map_element = baz
```

Since this behavior is confirmed fixed in the upcoming terraform 0.12 release, I will close this issue. Thanks again @jmoney8080.

**Comment by jmoney8080** [comment]
Do not know if it'll help with the locals trick as i'm trying to access map nested in a resource so the ordering of when locals and resources are processed may be an issue.  But glad to here v0.12 fixes it!  Two questions however
1) When is v0.12 released as I just see references that it's in beta?
2) any chance that it could be backported into 0.11.x given 0.12 is a major release?

**Comment by mildwonkey** [comment]
0.12 is still an alpha release, actually, and we don't have a firm timeline for the beta or general releases. I'm sorry that I can give a better answer - it's frustrating for all of us.

Unfortunately the configuration language enhancements are fundamental to the 0.12 release and cannot be back ported - [here](https://www.hashicorp.com/blog/terraform-0-12-rich-value-types) is the most relevant blog post regarding the enhanced configuration language.

Again, I am sorry that I cannot give a more exact (or satisfying) answer. 😞

**Comment by jmoney8080** [comment]
Just closing the loop here
```resource "aws_vpc_endpoint" "vpce" {
    vpc_id              = "${data.aws_security_group.internal.vpc_id}"
    service_name        = "${var.vpce_service_name}"
    vpc_endpoint_type   = "Interface"
    security_group_ids  = ["${data.aws_security_group.internal.id}"]
    subnet_ids          = ["${data.aws_subnet_ids.vpc_subnets.ids}"]
    private_dns_enabled = false
}


resource "aws_route53_record" "vpce-dns-record" {
    count = "${length(aws_vpc_endpoint.vpce.dns_entry)}"
    zone_id = "${var.r53_zone_id}"
    name    = "${var.application_service_name}.${var.aws_region}"
    type    = "CNAME"
    ttl     = "60"

    records = ["${lookup(aws_vpc_endpoint.vpce.dns_entry[count.index], "dns_name")}"]

    weighted_routing_policy {
        weight = 10
    }

    set_identifier = "vpce-endpoint-${count.index}"
}

**Comment by jmoney8080** [comment]
The above kind of works, the count is broken due to the dynamic count problem(which i've seen is also fixed in v0.12).  However if you hard code a static count number it works fine.  That static count just needs to be how many subnets to pull from the vpc that was used for the endpoint.

So, hooray for v0.12! Can't wait to make this all better when it's released :)

**Comment by gpanula** [comment]
@mildwonkey @apparentlymart 
> Since this behavior is confirmed fixed in the upcoming terraform 0.12 release, I will close this issue. 

Maybe instead of closing the tickets that are fixed in 0.12, they could be labeled/tagged with "fixed in 0.12" and then re-opened?

Then when 0.12 finally makes it out the door, all of the "fixed in 0.12" issues could be associated with that release and properly closed.

I'm stumbling into a fair number of issues that are closed with a "fixed in 0.12" but there isn't an ETA for 0.12.  

It becomes a bit frustrating to see an issued closed and think "excellent there is a solution to my problem" and then read thru discussion to find out the solution is _"wait for 0.12"_.
